src/test/scala/com/wavesplatform/history/BlockchainUpdaterKeyAndMicroBlockConflictTest.scala

```diff
@@ -0,0 +1,215 @@
+        secondAccount <- accountGen
+
+        tsAmount = FeeAmount * 10
+        transfer1 <- validTransferGen(richAccount, secondAccount, tsAmount)
+        transfer2 <- validTransferGen(secondAccount, richAccount, tsAmount - FeeAmount)
+        transfer3 <- validTransferGen(secondAccount, richAccount, tsAmount - FeeAmount)
+      } yield {
+        val genesisBlock = unsafeBlock(
+          reference = randomSig,
+          txs = Seq(GenesisTransaction.create(richAccount, tsAmount + FeeAmount, 0).explicitGet()),
+          signer = TestBlock.defaultSigner,
+          version = 3,
+          timestamp = 0
+        )
+
+        val (keyBlock, microBlocks) = unsafeChainBaseAndMicro(
+          totalRefTo = genesisBlock.signerData.signature,
+          base = Seq(transfer1),
+          micros = Seq(Seq(transfer2)),
+          signer = richAccount,
+          version = 3,
+          timestamp = System.currentTimeMillis()
+        )
+
+        val (keyBlock1, _) = unsafeChainBaseAndMicro(
+          totalRefTo = keyBlock.signerData.signature,
+          base = Seq(transfer3),
+          micros = Nil,
+          signer = secondAccount,
+          version = 3,
+          timestamp = System.currentTimeMillis()
+        )
+
+        (genesisBlock, keyBlock, microBlocks, keyBlock1)
+      }
+    }
+
+    def leaseAndLeaseCancel(): Gen[(Block, Block, Block, Seq[MicroBlock], Block, PrivateKeyAccount)] = {
+      for {
+        richAccount   <- accountGen
+        secondAccount <- accountGen
+        randomAccount <- accountGen
+
+        tsAmount = FeeAmount * 10
+        lease       <- validLeaseGen(richAccount, secondAccount, tsAmount)
+        leaseCancel <- validLeaseCancelGen(richAccount, lease.id())
+        transfer    <- validTransferGen(richAccount, randomAccount, tsAmount)
+      } yield {
+        val genesisBlock = unsafeBlock(
+          reference = randomSig,
+          txs = Seq(GenesisTransaction.create(richAccount, tsAmount + FeeAmount * 3, 0).explicitGet()),
+          signer = TestBlock.defaultSigner,
+          version = 3,
+          timestamp = 0
+        )
+
+        val leaseBlock = unsafeBlock(
+          genesisBlock.signerData.signature,
+          Seq(lease),
+          richAccount,
+          3,
+          System.currentTimeMillis()
+        )
+
+        val (keyBlock, microBlocks) = unsafeChainBaseAndMicro(
+          totalRefTo = leaseBlock.signerData.signature,
+          base = Nil,
+          micros = Seq(Seq(leaseCancel)),
+          signer = richAccount,
+          version = 3,
+          timestamp = System.currentTimeMillis()
+        )
+
+        val transferBlock = unsafeBlock(
+          keyBlock.signerData.signature,
+          Seq(transfer),
+          secondAccount,
+          3,
+          System.currentTimeMillis()
+        )
+
+        (genesisBlock, leaseBlock, keyBlock, microBlocks, transferBlock, secondAccount)
+      }
+    }
+
+    private[this] def unsafeChainBaseAndMicro(totalRefTo: ByteStr,
```

Comment:
These are copy-pasted, right? Could you move them somewhere, TestHelpers maybe?

Reply:
Ok